src/librustc_metadata/decoder.rs

```diff
@@ -824,12 +824,29 @@ impl<'a, 'tcx> CrateMetadata {
                         Def::Variant(def_id) => {
                             // Braced variants, unlike structs, generate unusable names in
                             // value namespace, they are reserved for possible future use.
-                            // It's ok to use the variant's id as a ctor id since an
-                            // error will be reported on any use of such resolution anyway.
-                            let ctor_def_id = self.get_ctor_def_id(child_index).unwrap_or(def_id);
+                            let (ctor_def_id, vis) = match self.get_ctor_def_id(child_index) {
+                                Some(ctor_def_id) =>
+                                    (ctor_def_id, self.get_visibility(ctor_def_id.index)),
```

Comment:
Why does `get_visibility` returns the correct (previously encoded) result for struct constructors above, but not for variant constructors here?

Reply:
So, when I remove this part then I get [this failure](https://gist.github.com/davidtwco/16210ca9f6d7d9912744385a89a56a12) (as well as a similar failure on a few other tests). This was one of the more painful issues I ran into when making this PR.

What seems to happen is that something in `rustc_resolve` fails when there is a struct variant defined in an external crate, as that crate won't have a ctor in the metadata. This results in a path being resolved as `QPath::TypeRelative` rather than `QPath::Resolved` which results in the error changing during type check. 

I spent quite a while trying to track down where it ended up being that a struct variant's ctor was expected but eventually landed on this as a fix. Unfortunately, I've forgotten the call stack I worked out was being taken.

Reply:
Hmm, I've just reverted this change and got no test failures.
Are you sure it still reproduces after the changes in https://github.com/rust-lang/rust/pull/59382?

Reply:
🤦‍♂️ I was mirroring the `Def::Struct(..)` arm instead of reverting. You're correct, there aren't any test failures. I've pushed that change.